tpm2/kdf.go

```diff
@@ -68,3 +68,43 @@ func KDFa(hashAlg Algorithm, key []byte, label string, contextU, contextV []byte
 
 	return out, nil
 }
+
+// KDFe implements TPM 2.0's ECDH key derivation function, as defined in
+// section 11.4.9.3 of the TPM revision 2 specification part 1.
+// See: https://trustedcomputinggroup.org/resource/tpm-library-specification/
+// The z parameter is the x coordinate of one parties private ECC key and the other parties public ECC key.
+// The partyUInfo and partyVInfo are the x coordinate of the initiators and the responders ECC points respectively.
+// Only the SHA256 hash algorithm is implemented at this time.
+func KDFe(hashAlg Algorithm, z []byte, use string, partyUInfo, partyVInfo []byte, bits int) ([]byte, error) {
+	var counter uint32
+	remaining := (bits + 7) / 8
+	var out []byte
+
+	var h hash.Hash
+	switch hashAlg {
+	case AlgSHA256:
+		h = sha256.New()
+	default:
+		return nil, fmt.Errorf("hash algorithm 0x%x is not supported", hashAlg)
+	}
+	for remaining > 0 {
+		counter++
+		if err := binary.Write(h, binary.BigEndian, counter); err != nil {
+			return nil, fmt.Errorf("pack counter: %v", err)
+		}
+		h.Write(z)
+		h.Write([]byte(use))
+		h.Write([]byte{0})
+		h.Write(partyUInfo)
+		h.Write(partyVInfo)
+		out = h.Sum(out)
+		remaining -= h.Size()
+		h.Reset()
+	}
+
+	if len(out) > bits/8 {
```

Comment:
Why do we round down here as opposed to rounding up above?

Reply:
This strips excess bytes if the expected number of bytes is not a multiple of the hash digest size.